### ISSUE TITLE
Performance regression since ol@6.13.0

### ISSUE THREAD

**Issue by iamvdo** [openlayers/openlayers#14151]
**Describe the bug**
In an app I'm working, I noticed a performance regression since ol@6.13.0.

The bug appears in a Vue@2 app. The map has one vector layer with interactions on it (modify, draw, snap). On a panel next to the map, there is a list of features, and hover over an item from the list update the style of the hovered feature. With ol@6.12.0, the style's update is very smooth, but with ol@6.13.0, it takes 1/2 seconds to update style (see videos). The app also tries to minimize vue impact, in a way that openlayers object like Map, VectorLayer, Interactions, Style are outside the vue app (well, not really outside, but vue reactivity is stopped).

<details><summary>Videos comparing perf regression between 6.12 and 6.13</summary>

https://user-images.githubusercontent.com/1260529/192971030-6c2adedb-51eb-42a2-8782-f3afdb73e3a8.mp4

https://user-images.githubusercontent.com/1260529/192971221-82a78f82-0a97-40c7-95f2-ec566e676f54.mp4
</details>

After investigating further, [this commit](https://github.com/openlayers/openlayers/commit/eb4d5e0784d347b044273cedb4a7a184d3c9192e) causes the issue, and specifically, [this line](https://github.com/openlayers/openlayers/blob/5e8223ddcb0b55c92345f86cd4418a0ec474fb0f/src/ol/renderer/canvas/Layer.js#L265)

https://github.com/openlayers/openlayers/blob/b06ffe5c97f8d98b6f4e973785b8b70f7ee658b7/src/ol/renderer/canvas/Layer.js#L265

So, my main question is: **what this line is all about?**

And second question: **do you have tips to fix the issue?**
I'm aware that the problem is maybe not specifically from openlayers, but it used to work well before ol@6.13.0

Also, if you need it, I can make a webpage with the reduced testcase to make your own tests.

Thanks.
Vincent

**Comment by M393** [comment]
Seems strange that this would be caused by #13338 as you are not reading pixel data.
How did you determine the issue is caused by that line? It is a simple assignment, shouldn't impact performance at all.

I am not familiar with Vue, but could it be doing some data binding magic? That likely is unwanted for the map.
See https://github.com/openlayers/openlayers/issues/13849#issuecomment-1187159264

**Comment by M393** [comment]
Have you tried to profile what functions are being called?
https://developer.chrome.com/docs/devtools/evaluate-performance/

**Comment by ahocevar** [comment]
@iamvdo When you register listeners listeners on OpenLayers Observables, do you store keys for use with `unByKey()`? That usually causes Vue data binding wrappers to kick in. The existence of Vue data bindings would also explain why that line you referenced above (`this.frameState = frameState`) makes such a performance difference.

**Comment by 1249993110** [comment]
Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently

**Comment by ahocevar** [comment]
@1249993110 That's correct, but does not apply here. However, we should be setting that attribute on the canvas we create for the `Layer.getFeatures(pixel)` method.

**Comment by 1249993110** [comment]
@ahocevar Thanks for your valuable answer.

**Comment by iamvdo** [comment]
@ahocevar No, I'm not storing keys, nor use `unByKey()`.
For now, it seems that I've found a workaround for the app. But still strange behaviour.

A Vue component adds reactivity to an array of ~10 OL layers (`ol/layer/Tile`, `ol/tilegrid/WMTS`), to create a layer selector (and update a visibility icon accordingly). Moved this bunch of layers outside of Vue's scope also fix the mentioned bug here. Weird, because nothing is related, and it worked fine before `ol@6.13`.

**Comment by iamvdo** [comment]
@ahocevar No, I'm not storing keys, nor use `unByKey()`.

I published the reduced testcase from the videos, so you can test for yourself.

1. Open https://codesandbox.io/s/naughty-chihiro-sd6jry
2. Click the "Load GPX" button (it will load a gpx and draw features on the map)
3. Hover over the tag labels (representing line chunks) on the left, it updates the feature's style (with 6.12.0, it is pretty smooth)
4. Update ol to 6.13.0 in `package.json`
5. Wait, and test again number 3, it is no longer smooth

Here is a list of important things to note:
- most of OL objects are outside Vue (simple variables)
- the map is built on the `created` hook (should not be reactive)
- the basemap layer is stored in a reactive key (not used here in this specific testcase)

Each workarounds I've found that fixes the "bug" (each one fixes the bug on its own):
- Do not add modify and snap interactions to the map (comment lines 162 and 172)
- Move the map creation outside Vue
- Remove the basemap layer from reactive data (comment line 95 `layer: basemapLayer`)

I spend many hours trying to debug this one, and building a reduce testcase that fits in one component. The real app has more components, and it's very hard sometimes to reason about. Also, some fixes won't work (eg, make the map as a simple variable)

The point is that Vue is clearly the culprit here, but it used to work well before `ol@6.13.0`
